### PR TITLE
[MRG] Implement c_code for GpuReshape

### PR DESCRIPTION
**Pull request by ejls**
WIP

### CODE REVIEW

File: theano/sandbox/gpuarray/basic_ops.py [hunk 1]
```diff
@@ -815,6 +815,44 @@ def perform(self, node, inp, out_):
...
+                                 compute_axis, i);
+                    %(fail)s;
+                }
+                compute_axis = i;
+            }
+        }
+        
+        Py_XDECREF(%(output)s);
+        %(output)s = pygpu_reshape(%(x)s, %(new_ndim)s, new_dims,
+                                   GA_ANY_ORDER, 1, compute_axis);
```
Comment (carriepl) [inline theano/sandbox/gpuarray/basic_ops.py]: What happens when the new shape does not contain the same number of elements as the original shape? Is thia already verified by pygpu_reshape() and a reasonable error message is raised or should this be handled in this op?

**Comment by carriepl** [comment]
Other than the point I mentioned about validating the total number of elements, this looks pretty good.

**Comment by nouiz** [comment]
pygpu_reshape probably give an unclear error. But you can test it to make
sure.

**Comment by ejls** [comment]
I've added a commit addressing this, if it's fine as is, I'll fusion the commits into one.

File: theano/sandbox/cuda/basic_ops.py [hunk 2]
```diff
@@ -2437,6 +2437,88 @@ def perform(self, node, inp, out_):
...
+            PyObject *shape_to_py = PyObject_Str(new_shape);
+            const char *shape_to = PyString_AsString(shape_to_py);
+            PyErr_Format(PyExc_ValueError,
+                         "GpuReshape: cannot reshape input of shape "
+                         "%%s to shape %%s.", shape_from, shape_to);
+            %(fail)s;
+        }
+
+        Py_XDECREF(%(output)s);
+        %(output)s = (CudaNdarray*) CudaNdarray_Reshape(%(x)s, new_shape);
```
Comment (nouiz) [inline theano/sandbox/cuda/basic_ops.py]: If output is NULL, then exec fail.

File: theano/sandbox/cuda/basic_ops.py [hunk 3]
```diff
@@ -2437,6 +2437,88 @@ def perform(self, node, inp, out_):
...
+        output, = outputs
+        new_ndim = self.ndim
+        sdtype = node.inputs[1].type.dtype_specs()[1]
+        fail = sub['fail']
+        return """
+        assert (PyArray_NDIM(%(shape)s) == 1);
+        PyObject *new_shape = PyTuple_New(%(new_ndim)s);
+        int compute_axis = -1;
+        size_t total = 1;
+        for (size_t i = 0; i < %(new_ndim)s; ++i)
```
Comment (nouiz) [inline theano/sandbox/cuda/basic_ops.py]: check that the lenght of shape is new_ndim.

Comment (nouiz) [inline theano/sandbox/cuda/basic_ops.py]: idem for gpuarray back-end

File: theano/sandbox/cuda/basic_ops.py (hunk 2 above)
Comment (nouiz) [inline theano/sandbox/cuda/basic_ops.py]: idem for gpuarray.

**Comment by nouiz** [comment]
except for the smalll comment, it look good.

**Comment by ejls** [comment]
Thanks for the review, I've updated the PR.

**Comment by carriepl** [comment]
This looks good. I am locally running the GPU tests for both backends with this PR. If they pass, it should be fine for merge.

**Comment by carriepl** [comment]
On the CUDA backend, only one test fails (test_memory.py) but it was already failing before any commit in this pull request. @nouiz is currently running the tests on the Gpuarray backend because I seem to have to issues with my installation.

**Comment by nouiz** [comment]
I have this error in the new gpu back-end:
```
======================================================================
FAIL: test_reshape (theano.sandbox.gpuarray.tests.test_basic_ops.G_reshape)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/u/bastienf/repos/Theano.Clean/theano/tensor/tests/test_basic.py", line 4913, in test_reshape
    assert numpy.all(f_out2 == c_val2), (f_out2, c_val2)
AssertionError: (array([ 0.,  1.,  2.,  3.,  4.,  5.]), array([0, 3, 1, 4, 2, 5]))

----------------------------------------------------------------------

```

**Comment by ejls** [comment]
I've squashed the commits together.
I see the Travis build failed but it seems to come from somewhere else.

**Comment by carriepl** [comment]
Have you corrected the failing test under the new gpu backend? For Travis, I'm trying to see what test failed but all I'm getting is a blank page for now.

**Comment by ejls** [comment]
On Fri, Jun 19, 2015 at 05:16:10AM -0700, carriepl wrote:
> Have you corrected the failing test under the new gpu backend? For Travis, I'm trying to see what test failed but all I'm getting is a blank page for now.
Yes, sorry I didn't comment between committing the fix and squashing the commits.
I had misunderstood the parameters of pygpu_reshape, it's the only line I changed.

-- 
Étienne Simon

**Comment by nouiz** [comment]
closed/reopened to rerun travis with new master.

**Comment by carriepl** [comment]
The tests are passing for the old gpu backend.

**Comment by nouiz** [comment]
It passed. We can merge when the python2/3 PR is merged thanks.

File: theano/sandbox/cuda/basic_ops.py [hunk 4]
```diff
@@ -2437,6 +2437,102 @@ def perform(self, node, inp, out_):
...
+        return (1,)
+
+    def c_code(self, node, name, inputs, outputs, sub):
+        x, shape = inputs
+        output, = outputs
+        new_ndim = self.ndim
+        sdtype = node.inputs[1].type.dtype_specs()[1]
+        fail = sub['fail']
+        return """
+        PyObject *new_shape = PyTuple_New(%(new_ndim)s);
```
Comment (nouiz) [inline theano/sandbox/cuda/basic_ops.py]: Just to let know this object wasn't freed. So there was a memory leak on the CPU. This is fixed by gh-3193

**Pull request merged**